### ISSUE TITLE
Support Async APIs for Lambda RequestHandler Interface

### ISSUE THREAD

**Issue by bytekast** [aws/aws-sdk-java#1164]
perhaps something like
```
public interface RequestHandler<Future<I>, Future<O>> {
  public Future<O> handleRequest(Future<I> input, Context context);
}
```

**Comment by shorea** [comment]
Apologies for the delay. This seems to be a feature request for the actual lambda handler, is that a correct assumption? I.E an addition to http://docs.aws.amazon.com/lambda/latest/dg/java-programming-model-handler-types.html

**Comment by bytekast** [comment]
Yes - An asynchronous Lambda Handler for Java, where the handler can return
a Future/Promise object.

- Rowell

**Comment by shorea** [comment]
Gotcha, we don't control the handler side of things. That's managed by the Lambda service team. I can pass this feedback along to them. Could you provide more details on your use case and why returning a future is desirable, I'm having trouble wrapping my head on how that will work with the lambda container architecture. Any details you can provide will help the service team with prioritization.

**Comment by bytekast** [comment]
@shorea its essentially the same async callback mechanism provided in the JavaScript handler. Many JVM frameworks/applications now support async handlers. It would be nice if the AWS Lambda Java SDK supports this programming model as well.

**Comment by pendula95** [comment]
Is there any progress on this? Example: Vertx framework. If I want to implement custom API gateway that would be invoked from http request, I need a async http request handler for lambda as vertx is a async framework. As @shorea mention this is already supported in AWS JavaScript handler.

**Comment by dagnir** [comment]
@pendula95 Have you opened a ticket through the [AWS Support Center](https://console.aws.amazon.com/support/)? It might be a more expedient approach since you'll have a more direct line of communication with Lambda that way.

**Comment by pendula95** [comment]
I am only testing features to see if AWS is suitable for our use case. I only have Basic Support Plan so I can't contact technical support.

**Comment by RichardSilveira** [comment]
+1 (additionally on it, it´s normal return a 202 response on REST API´s...)

**Comment by muhammadssr** [comment]
Nothing on this yet?

**Comment by debora-ito** [comment]
Echoing comments in the issue submitted in https://github.com/aws/aws-lambda-java-libs/issues/44, there's no point for a handler to be async because "if a second invoke of a Lambda function is made before the first has finished a new execution environment will be created".

Closing this.
